pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -0,0 +1,220 @@
+package sysregistriesv2
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/BurntSushi/toml"
+	"github.com/containers/image/types"
+	"github.com/sirupsen/logrus"
+	"github.com/containers/image/docker/reference"
+
+	"io/ioutil"
+	"path/filepath"
+)
+
+// systemRegistriesConfPath is the path to the system-wide registry configuration file
+// and is used to add/subtract potential registries for obtaining images.
+// You can override this at build time with
+// -ldflags '-X github.com/containers/image/sysregistries.systemRegistriesConfPath=$your_path'
+var systemRegistriesConfPath = builtinRegistriesConfPath
+
+// builtinRegistriesConfPath is the path to registry configuration file
+// DO NOT change this, instead see systemRegistriesConfPath above.
+const builtinRegistriesConfPath = "/etc/containers/registries.conf"
+
+// RegURL represents a Registry's URl.  This abstraction is required to
+// unmarshal the non-primitive and non-local url.URL type when loading the
+// toml config.
+type RegURL struct {
+	URL url.URL
+}
+
+// String returns host:port/path of the RegURL's URL.
+func (r *RegURL) String() string {
+	return r.URL.Host + r.URL.Path
+}
+
+// newRegURL parses the input string, performs some sanity checks and returns
+// a url.URL.  The input must be a valid URI with an "http" or "https" scheme,
+// a specified host and an empty URI user.  Otherwise, an error is returned.
+func newRegURL(input string) (url.URL, error) {
+	uri, err := url.Parse(input)
+	if err != nil {
+		return url.URL{}, fmt.Errorf("error parsing URL %s: %v", input, err)
+	}
+
+	// only https and http are valid URI schemes
+	if uri.Scheme == "" {
+		return url.URL{}, fmt.Errorf("unspecified URI scheme: %s", input)
+	}
+	if uri.Scheme != "https" && uri.Scheme != "http" {
+		return url.URL{}, fmt.Errorf("unsupported URI scheme: %s", uri.Scheme, input)
+	}
+
+	// a host must be specified
+	if uri.Host == "" {
+		return url.URL{}, fmt.Errorf("unspecified URI host: %s", input)
+	}
+
+	// URI user must be empty
+	if uri.User != nil {
+		// strip password for security reasons
+		uri.User = url.UserPassword(uri.User.Username(), "xxxxxx")
+		return url.URL{}, fmt.Errorf("unsupported username/password: %q", uri)
+	}
+
+	return *uri, nil
+}
+
+// UnmashalText turns text into a RegURL, and is a way to unmarshal a string
+// into a url.URL.
+func (r *RegURL) UnmarshalText(text []byte) (err error) {
+	r.URL, err = newRegURL(string(text))
+	return err
+}
+
+// Registry is a generalized type representing a registry host and a registry
+// mirror, whereas the mirror cannot further be mirrored.
+type Registry struct {
+	URL      RegURL     `toml:"url"`
+	Blocked  bool       `toml:"blocked"`
+	Insecure bool       `toml:"insecure"`
+	Search   bool       `toml:"unqualified-search"`
+	Prefix   string     `toml:"prefix"`
```

Comment:
What does this mean/do? Please add documentation to the field.

(And about `url = "foo"; prefix = "bar"`? Is that allowed? Reasonable?)

Reply:
I will add documentation. Sorry, in case that's confusing. Some discussion happened in IRC.

This is something similar to the moby/moby PR. The Prefix effectively allows to translate one namespace into another. So `url = "foo"; prefix = "bar";` would be allowed.

Notice that specifying a prefix is *not* mandatory.